### ISSUE TITLE
Humidity Reporting Fine, Temperature "Unknown" on LYWSD03MMC (ATC_MiThermometer)

### ISSUE THREAD

**Issue by anndrewww** [custom-components/sensor.mitemp_bt#132]
Two (2) LYWSD03MMC devices flashed with the ATC_MiThermometer .bin.

Both are reporting Humidity with no issues but Temperature on both is reporting as "Unknown".

Both have the following settings via the Web Flasher:

> Smiley: Off 
> 
> Advertising Type:  Mi Like
> 
> Sensor display: In °F
> 
> Show battery in LCD: Disabled
> 
> Advertising interval: 1 Minute

My Mi_Temp.yaml is:

> platform: mitemp_bt
>     rounding: True
>     decimals: 1
>     period: 60
>     log_spikes: False
>     use_median: False
>     active_scan: False
>     hci_interface: 0
>     batt_entities: False

**Comment by Ernst79** [comment]
My first guess, perhaps changing the sensor display to Fahrenheit also changes the ble messages to Fahrenheit.  

20 degrees Celsius will become 68 degrees Fahrenheit, which is outside the range as set in const.py. Could you try to change the following two lines in const.py to appropriate values in Fahrenheit?

CONF_TMIN = -40.0
CONF_TMAX = 60.0

Please restart HA afterwards and report back your findings.

**Comment by anndrewww** [comment]
I changed the display back to Celsius and the values started to report and convert to Fahrenheit for display into HA. 

If I edit Const.py, values start reporting to HA but are nearly doubled because it's treating them as Celisus and applying the conversion.

**Comment by zanchin** [comment]
Same observations here. When I set the display to Fahrenheit and change const.py, I see correct entity attributes:

```yaml
last packet id: 116
rssi: -66
sensor type: LYWSD03MMC
mac address: 'A4:C1:38:B8:30:2D'
battery_level: 85
last mean of: 4
median: 74.6
mean: 74.6
unit_of_measurement: °F
friendly_name: mi temperature A4C138B8302D
device_class: temperature
```

However, when graphed, HA interprets it as a Celsius temperature and multiplies the value (the jump is when I switched from C to F):
![image](https://user-images.githubusercontent.com/281978/97323343-1771a300-1847-11eb-925d-c6ba0afec808.png)

Note than when display was set to C, HA was graphing temperature with the F value, but with the °C unit, which means it can't be graphed next to other °F sensors.

Is this controllable from within the integration, or is it a setting in HA?

**Comment by Ernst79** [comment]
Ok, let me summarize to understand what is happening. Range is currently set as a fixed value in const.py:

Display in Celsius --> within range -40°C - +60°C --> unit_of measurement = °C --> Displayed correctly in HA in °C
Display in Celsius --> within range -40°C - +60°C --> **unit_of measurement = ??** --> Displayed correctly in HA in °F
Display in Fahrenheit --> outside range -40°F - +60°F --> Unknown (makes sense, as it is outside the range)
Display in Fahrenheit --> within range -40°F - +160°F --> unit_of measurement = °F --> Displayed incorrect in HA (converted twice to Fahrenheit, see picture above)

**Comment by anndrewww** [comment]
From the HomeAssistant Docs:

> As part of the default onboarding process, Home Assistant can detect your location from IP address geolocation. Home Assistant will automatically select a temperature unit and time zone based on this location. You may adjust this during onboarding, or afterwards at Configuration -> General.

> If you prefer YAML, you can add the following information to your configuration.yaml:

> unit_system string (optional)
> metric for Metric, imperial for Imperial. This also sets temperature_unit, Celsius for Metric and Fahrenheit for Imperial
> 
> temperature_unit string (optional)
> Override temperature unit set by unit_system. C for Celsius, F for Fahrenheit.

For reference, my HA setup is set via "Configuratino -> General" to "Imperial"

**Comment by zanchin** [comment]
@Ernst79 That's a correct summary. For the second line, unit_of measurement is set to F globally in my HA (as the Unit System). So a configuration option to expect temperature in F makes sense. It would also make sense for the sensor itself show the temp in F but broadcast in C (this would be a firmware change).

As for my point about F values displayed in C, I figured that out. My initial data had `unit_of measurement = °C` in the database, so the graph was using the first datapoint as the unit. Once I deleted all of those entries (see [here](https://community.home-assistant.io/t/delete-history-of-one-sensor/87203/8)) the graph correctly displayed  data with the °F unit.

Now that I have the device show the temp in C (and broadcast in C), everything renders correctly in HA.

So the only real issue is when the device is set to display in F (it broadcasts in F).

**Comment by Ernst79** [comment]
Thanks for confirming. I'm working on it, but it can take a couple of days, as I have to do it in my spare time. 

Just for my understanding, the sensor with original firmware, can this also be set to °F and if so, was it reporting OK in Home Assistant? In other words, is this only applicable for sensors with ATC_MiThermometer.bin firmware?

As you mention, there are two solutions, making the firmware sending in °C (with display in F) or telling mitemp_bt that it is getting data in °F. I can only do the last option, as I'm not the developer of the firmware. So, I'll work on the second option. 

Progress
- [x] Configuration option 
- [ ] Setting correct unit_of_measurement (depending on config)
- [ ] Modify sensor limits to values in °F (depening on config)

**Comment by zanchin** [comment]
I never ran the sensor with the original firmware, only the custom one.

I can help with the patch if you point me to the right placed in the code to look at. I assume correct setting of `unit_of_measurement` would be somewhere around where `TEMP_CELSIUS` is set?

**Comment by Ernst79** [comment]
I will upload the current state later, have it locally. Have finished the unit_of_measurement, already. Modifying the sensor limits is the next thing.

**Comment by anndrewww** [comment]
I can confirm this (single sensor) formatting is working:

`sensor_fahrenheit: 'XX:XX:XX:XX:XX:XX'`

This (multiple sensor) formatting is NOT working:

```
  sensor_fahrenheit:
    'XX:XX:XX:XX:XX:XX'
    'XX:XX:XX:XX:XX:XX'
```
Error in Check Config:
```
                          Error loading /config/configuration.yaml: while parsing a block mapping
  in "/config/sensors/mi_temp.yaml", line XX, column XX
expected <block end>, but found '<scalar>'
  in "/config/sensors/mi_temp.yaml", line XX, column XX
```

**Comment by Ernst79** [comment]
Ah, sorry, we have to use the same format a the whitelist. `-` was missing.

```
  sensor_fahrenheit:
    - 'C4:7C:8D:6B:4F:F3'
    - '4C:65:A8:DD:B8:9B'
```

**Comment by anndrewww** [comment]
Updated - Now displaying/reporting correctly in HA with two (2) sensors. Both custom firmware, with "Mi Like Advertising", and Fahrenheit displayed. Other older model (non-custom firmware) sensors are working as expected and don't appear to be effected by the change.

**Comment by zanchin** [comment]
I only have one sensor at the moment, but it's working great now as well. Thank you!

**Comment by Ernst79** [comment]
Ok, thanks for reporting back. I have to adjust the sensor limits, which are set in the code, to be depending on Fahrenheit/Celsius. But for now, you can continue to use it. After the release, you can probably just update and it will continue to work.

**Comment by anndrewww** [comment]
Great! Thanks, @Ernst79

**Comment by Ernst79** [comment]
Please keep the issue open till it is released in HACS. I will close it later

**Comment by Ernst79** [comment]
I've released 0.7.7-beta, including the last fix for the sensor limits. Please try (you can update via HACS, by allowing beta releases). configuration remains the same.
